bindings/node/test/clientEncryption.test.js

```diff
@@ -0,0 +1,139 @@
+const MongoClient = require('mongodb').MongoClient;
+const fs = require('fs');
+const expect = require('chai').expect;
+const sinon = require('sinon');
+const StateMachine = require('../lib/stateMachine').StateMachine;
+const SegfaultHandler = require('segfault-handler');
+SegfaultHandler.registerHandler();
+
+function readHttpResponse(path) {
+  let data = fs.readFileSync(path, 'utf8').toString();
+  data = data.split('\n').join('\r\n');
+  return Buffer.from(data, 'utf8');
+}
+
+const ClientEncryption = require('..').ClientEncryption;
+describe('ClientEncryption', function() {
+  let client;
+  let sandbox = sinon.createSandbox();
+
+  after(() => sandbox.restore());
+  before(() => {
+    // stubbed out for AWS unit testing below
+    const MOCK_KMS_ENCRYPT_REPLY = readHttpResponse(`${__dirname}/data/kms-encrypt-reply.txt`);
+    sandbox.stub(StateMachine.prototype, 'kmsRequest').callsFake(request => {
+      request.addResponse(MOCK_KMS_ENCRYPT_REPLY);
+      return Promise.resolve();
+    });
+  });
+
+  beforeEach(() => {
+    client = new MongoClient('mongodb://localhost:27017/test', { useNewUrlParser: true });
+    return client.connect();
+  });
+
+  afterEach(() => {
+    return client.close();
+  });
+
+  [
+    {
+      name: 'local',
+      kmsProviders: { local: { key: Buffer.alloc(96) } }
+    },
+    {
+      name: 'aws',
+      kmsProviders: { aws: { accessKeyId: 'example', secretAccessKey: 'example' } },
+      options: { masterKey: { region: 'region', key: 'cmk' } }
+    }
+  ].forEach(providerTest => {
+    it(`should create a data key with the "${providerTest.name}" KMS provider`, function(done) {
+      const providerName = providerTest.name;
+      const encryption = new ClientEncryption(client, {
+        keyVaultNamespace: 'client.encryption',
+        kmsProviders: providerTest.kmsProviders
+      });
+
+      const dataKeyOptions = providerTest.options || {};
+      encryption.createDataKey(providerName, dataKeyOptions, (err, dataKey) => {
+        expect(err).to.not.exist;
+        expect(dataKey._bsontype).to.equal('Binary');
+
+        client
+          .db('client')
+          .collection('encryption')
+          .findOne({ _id: dataKey }, (err, doc) => {
+            expect(err).to.not.exist;
+            expect(doc).to.have.property('masterKey');
+            expect(doc.masterKey).to.have.property('provider');
+            expect(doc.masterKey.provider).to.eql(providerName);
+            done();
+          });
+      });
+    });
+  });
+
+  it('should explicitly encrypt and decrypt with the "local" KMS provider', function(done) {
+    const encryption = new ClientEncryption(client, {
+      keyVaultNamespace: 'client.encryption',
+      kmsProviders: { local: { key: Buffer.alloc(96) } }
+    });
+
+    encryption.createDataKey('local', (err, dataKey) => {
+      expect(err).to.not.exist;
+
+      const encryptOptions = {
+        keyId: dataKey,
+        algorithm: 'AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic'
```

Comment:
Is this a constant somewhere?

Reply:
it's not, but I'm not sure what the constant would be, I guess just the full string? This is a string meant to be passed to libmongocrypt, it will validate in the library